update.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import os
+from datetime import datetime
+from jinja2 import Environment, FileSystemLoader
+from typing import NamedTuple
+from urllib.error import URLError
+from urllib.request import urlopen, Request
+
+JDK_URLS = {
+    (12, 0, 1): 'https://download.java.net/java/GA/jdk12.0.1/69cfe15208a647278a19ef0990eea691/12/GPL/openjdk-12.0.1_linux-x64_bin.tar.gz',
+    (11, 0, 1): 'https://download.java.net/java/GA/jdk11/13/GPL/openjdk-11.0.1_linux-x64_bin.tar.gz'
+}
+
+
+class Version(NamedTuple):
+    major: int
+    minor: int
+    hotfix: int
+
+    @classmethod
+    def parse(cls, s: str):
+        return Version(*map(int, s.split('.', maxsplit=2)))
+
+    def __str__(self) -> str:
+        return f'{self.major}.{self.minor}.{self.hotfix}'
+
+
+def latest_crash():
+    with urlopen('https://crate.io/versions.json') as r:
+        d = json.load(r)
+        return Version.parse(d['crash'])
+
+
+def jdk_url_and_sha(jdk_version):
+    if jdk_version not in JDK_URLS:
+        raise ValueError(f'No URL for JDK version {jdk_version} found.')
+    url = JDK_URLS[jdk_version]
+    with urlopen(url + '.sha256') as r:
+        sha256 = r.read().decode('utf-8')
+    return url, sha256
+
+
+def url_exists(url: str):
+    try:
+        with urlopen(Request(url, method='HEAD')):
+            return True
+    except URLError:
+        return False
+
+
+def ensure_existing_crash(crash_version: Version):
+    if not crash_version:
+        return latest_crash()
```

Comment:
i am wondering whether we need this fallback, shouldn't we be strict about version we provide and fail fast

Reply:
This is currently done in the Jenkins job we have and I want to move the logic out of it so that it's easier to test things locally.